sdk/clientcore/core/src/main/java/io/clientcore/core/models/CoreException.java

```diff
@@ -0,0 +1,120 @@
+     */
+    public static CoreException from(Throwable cause) {
+        return from(null, cause);
+    }
+
+    /**
+     * Translates a {@link Throwable} into a {@link CoreException}.
+     *
+     * @param message the message to use instead of the cause's message
+     * @param cause the {@link Throwable} to translate
+     * @return the {@link CoreException} that was created
+     */
+    public static CoreException from(String message, Throwable cause) {
+        return from(message, cause, RetryUtils.isRetryable(cause));
+    }
+
+    /**
+     * Translates a {@link Throwable} into a {@link CoreException}.
+     *
+     * @param cause the {@link Throwable} to translate
+     * @param isRetryable whether the exception is retryable. When in doubt, set to {@code true}.
+     * @return the {@link CoreException} that was created
+     */
+    public static CoreException from(Throwable cause, boolean isRetryable) {
+        return from(null, cause, isRetryable);
+    }
+
+    /**
+     * Translates a {@link Throwable} into a {@link CoreException}.
+     *
+     * @param message the message to use instead of the cause's message
+     * @param cause the {@link Throwable} to translate
+     * @param isRetryable whether the exception is retryable. When in doubt, set to {@code true}.
+     * @return the {@link CoreException} that was created
+     */
+    static CoreException from(String message, Throwable cause, boolean isRetryable) {
+        String updatedMessage = message;
+        Throwable updatedCause = cause;
+        if (cause instanceof CoreException) {
+            CoreException e = (CoreException) cause;
+            if (e.isRetryable() == isRetryable && message == null) {
+                return e;
+            }
+
+            updatedMessage = getMessage(message, cause);
+            updatedCause = cause.getCause();
+        } else if (cause instanceof UncheckedIOException) {
+            updatedMessage = getMessage(message, cause);
+            updatedCause = cause.getCause();
+        }
+
+        return new CoreExceptionImpl(updatedMessage, updatedCause, isRetryable);
+    }
+
+    /**
+     * A boolean indicating whether the exception is retryable.
+     */
+    private final boolean isRetryable;
+
+    /**
+     * Creates a new instance of {@link CoreException}.
+     *
+     * @param message The exception message.
+     * @param cause The {@link Throwable} which caused the creation of this CoreException.
+     * @param isRetryable Whether the exception is retryable. When in doubt, set to {@code true}.
+     */
+    protected CoreException(String message, Throwable cause, boolean isRetryable) {
+        super(getMessage(message, cause), cause);
+        this.isRetryable = isRetryable;
+    }
+
+    /**
+     * Gets whether the exception is retryable.
+     *
+     * @return {@code true} if the exception is retryable; {@code false} otherwise.
+     */
+    public boolean isRetryable() {
```

Comment:
Since this is an abstract class now, should we make this an abstract method so classes extending from it will be required to implement this method?